llama_index/query_engine/sql_vector_query_engine.py

```diff
@@ -0,0 +1,311 @@
+        # first query SQL database
+        sql_response = self._sql_query_tool.query_engine.query(query_bundle)
+        if not self._use_sql_vector_synthesis:
+            return sql_response
+
+        sql_query = (
+            sql_response.extra_info["sql_query"] if sql_response.extra_info else None
+        )
+        if self._verbose:
+            print_text(f"SQL query: {sql_query}\n", color="yellow")
+            print_text(f"SQL response: {sql_response}\n", color="yellow")
+
+        # given SQL db, transform query into new query
+        new_query = self._sql_augment_query_transform(
+            query_bundle.query_str,
+            extra_info={
+                "sql_query": _format_sql_query(sql_query),
+                "sql_query_response": str(sql_response),
+            },
+        )
+
+        if self._verbose:
+            print_text(
+                f"Transformed query given SQL response: {new_query.query_str}\n",
+                color="blue",
+            )
+        logger.info(f"> Transformed query given SQL response: {new_query.query_str}")
+        if self._sql_augment_query_transform.check_stop(new_query):
+            return sql_response
+
+        vector_response = self._vector_query_tool.query_engine.query(new_query)
+        if self._verbose:
+            print_text(f"Vector DB response: {vector_response}\n", color="pink")
+        logger.info(f"> Vector DB response: {vector_response}")
+
+        response_str, _ = self._service_context.llm_predictor.predict(
+            self._sql_vector_synthesis_prompt,
+            query_str=query_bundle.query_str,
+            sql_query_str=sql_query,
+            sql_response_str=str(sql_response),
+            vector_store_query_str=new_query.query_str,
+            vector_store_response_str=str(vector_response),
+        )
+        if self._verbose:
+            print_text(f"Final response: {response_str}\n", color="green")
+        response_extra_info = {
+            **(sql_response.extra_info or {}),
+            **(vector_response.extra_info or {}),
+        }
+        source_nodes = vector_response.source_nodes
+        return Response(
+            response_str,
+            extra_info=response_extra_info,
+            source_nodes=source_nodes,
+        )
+
+    def _query(self, query_bundle: QueryBundle) -> RESPONSE_TYPE:
+        """Query and get response."""
+        # TODO: see if this can be consolidated with logic in RouterQueryEngine
+        metadatas = [self._sql_query_tool.metadata, self._vector_query_tool.metadata]
+        result = self._selector.select(metadatas, query_bundle)
+        # pick sql query
+        if result.ind == 0:
+            if self._verbose:
+                print_text(f"Querying SQL database: {result.reason}\n", color="blue")
+            logger.info(f"> Querying SQL database: {result.reason}")
+            return self._query_sql_vector(query_bundle)
+        elif result.ind == 1:
```

Comment:
So if it decides to query vector store, then don't need to run SQL query?

Reply:
Kinda unintuitive to me why it's asymmetric (SQL and vector store query)

Reply:
I could imagine cases where you'd want to go the other way (e.g. vector first, then SQL)

Reply:
Mm yeah more like I haven't really scoped out what that would mean - do you have an example use case in mind?